tensor_comprehensions/library/layers.yaml

```diff
@@ -0,0 +1,247 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+##############################################################################
+
+# Purpose:
+# This module serves as a database for Tensor Comprehensions representation of layers
+# that are useful for the Machine Learning community. This database should be
+# extended with more layers including fusions, compositions etc.
+#
+# To extend this:
+# add a new entry which at least has the "name" and the "lang". You can chose to
+# add other entries like "grad".
+# For the gradient layers, name should be: name + "_grad"
+
+- name: indexing
+  lang: |
+    def indexing(float(H, W) input, int32(L) index) -> (output) {{
+      output(l, w) = input(index(l), w) where l in 0:{L}
+    }}
+
+- name: lookup_table
+  lang: |
+    def lookup_table(float(B, R) LUT, int32(B, N) I) -> (O) {
+      O(b, n) +=! LUT(I(b, n), r)
+    }
+
+- name: matmul
+  lang: |
+    def matmul(float(M,N) A, float(N,K) B) -> (output) {
+      output(i, j) +=! A(i, kk) * B(kk, j)
+    }
+
+- name: batch_matmul
+  lang: |
+    def batch_matmul(float(B, N, M) X, float(B, M, K) Y) -> (Z) {
+      Z(b, n, k) +=! X(b, n, mm) * Y(b, mm, k)
+    }
+
+- name: transpose
+  lang: |
+    def transpose(float(N, C, H, W) I) -> (O) {
+      O(c, n, w, h) = I(n, c, h, w)
+    }
+
+- name: avgpool
+  lang: |
+    def avgpool(float(B, C, H, W) input) -> (output) {{
+      output(b, c, h, w) += input(b, c, h * {sH} + kh, w * {sW} + kw) where kh in 0:{kH}, kw in 0:{kW}
+    }}
+
+- name: maxpool
+  lang: |
+    def maxpool(float(B, C, H, W) input) -> (output) {{
+      output(b, c, h, w) max= input(b, c, h * {sH} + kh, w * {sW} + kw) where kh in 0:{kH}, kw in 0:{kW}
+    }}
+
+- name: scale
+  lang: |
+    def scale(float(M, N) I) -> (O) {{
+      O(m, n) = I(m, n) * {s}
+    }}
+
+- name: sigmoid
+  lang: |
+    def sigmoid(float(N, C, H, W) I) -> (O) {
+      O(n, c, h, w) = 1 / (1 + exp(-I(n, c, h, w)))
+    }
+
+- name: softmax
+  lang: |
+    def softmax(float(N, D) I) -> (O, expsum, maxVal) {
+      maxVal(n) max= I(n, d)
+      expsum(n) +=! exp(I(n, d) - maxVal(n))
+      O(n, d) = exp(I(n, d)) / expsum(n)
+    }
+
+- name: Tanh
+  lang: |
+    def Tanh(float(M) I) -> (O) {
+      O(m) = tanh(I(m))
+    }
+
+- name: tensordot
+  lang: |
+    def tensordot(float(N, C1, C2, H, W) I0, float(N, C2, C3, H, W) I1) -> (O) {
+      O(n, c1, c3, h, w) +=! I0(n, c1, c2, h, w) * I1(n, c2, c3, h, w)
+    }
+
+- name: fully_connected
+  lang: |
+    def fully_connected(float(B,M) I, float(N,M) W1, float(N) B1) -> (O1) {
```

Comment:
nit: spacings are inconsistent in the parameters in this def and a few others

Reply:
thanks for the nits, I'll correct them